lib/vwo/core/variation_decider.rb

```diff
@@ -178,6 +224,90 @@ def get_variation_allotted(user_id, campaign)
+      # @param[String]  :user_id      The unique key assigned to User
+      # @param[Hash]    :campaign     Campaign hash for Unique campaign key
+      #
+      # @return[Hash]
+
+      def evaluate_Whitelisting(user_id, campaign, campaign_key, variation_targeting_variables )
+        variation_targeting_variables = variation_targeting_variables.merge({ vwo_user_id: user_id })
+        targeted_variations = []
+
+        campaign['variations'].each do |variation|
+          segments = get_segments(variation)
+          is_valid_segments = valid_value?(segments)
+          if is_valid_segments
+            if @segment_evaluator.evaluate(campaign_key, user_id, segments, variation_targeting_variables)
+              targeted_variations.push(variation)
+            end
+          else
+            @logger.log(
+              LogLevelEnum::INFO,
+              format(
+                LogMessageEnum::InfoMessages::SKIPPING_PRE_SEGMENTATION,
+                file: FILE,
+                campaign_key: campaign_key,
+                user_id: user_id,
+                api_name: ApiMethods::GET_FEATURE_VARIABLE_VALUE,
+                varaition: variation['name']
+              )
+            )
+          end
+        end
+
+        if targeted_variations.length() > 1
+          scale_variation_weights(targeted_variations)
+          current_allocation = 0
+          targeted_variations.each do |variation|
+            step_factor = get_variation_bucketing_range(variation['weight'])
+            if step_factor > 0
+              start_range = current_allocation + 1
+              end_range = current_allocation + step_factor
+              variation['start_variation_allocation'] = start_range
+              variation['end_variation_allocation'] = end_range
+              current_allocation += step_factor
+            else
+              variation['start_variation_allocation'] = -1
+              variation['end_variation_allocation'] = -1
+            end
+          end
+          whitelisted_variation = @bucketer.get_variation(
+            targeted_variations,
+            @bucketer.get_bucket_value_for_user(
+              user_id
+            )
+          )
+        else
+          whitelisted_variation = targeted_variations[0]
+        end
+        whitelisted_variation
+      end
+
+      private
+
+      # It extracts the weights from all the variations inside the campaign
+      # and scales them so that the total sum of eligible variations' weights become 100%
+      #
+      # 1. variations
+
+      def scale_variation_weights(variations)
+        total_weight = variations.reduce(0) { |acc, variation| acc + variation['weight'] }
+        unless total_weight
+          weight = 100 / variations.length()
+          variations.each do |variation|
+            variation['weight'] = weight
+          end
+          return
+        end
+        variations.each do |variation|
+            variation['weight'] = (variation['weight'] / totalWeight) * 100
```

Comment:
no variable called `totalWeight`

Looks like no testing done and no rubocop executed.

Reply:
Fixed.